Support custom s3 api endpoints

easel opened the pull request:
Rebased #619 against latest master.

easel commented:
I've added some documentation and adjusted the s3 integration tests to work against Minio. Aside from this patch, getting minio to pass the tests requires falling back to url encoding the responses from the upload calls. 

The regular test suite still passes for me, and unless I'm missing something I don't believe any of the changes should change the code path unless an endpoint url is specified. Happy to address any other feedback.

@takashima0411 @ktoso any thoughts?

ktoso commented:
Sorry for the delay, family time back home until New Years :)
Will review then!

easel commented:
Thanks @ktoso, happy New Year!

ennru commented:
Does this replace #619 ? Please close it if so.

easel commented:
Yes, it's includes all of #619 but includes fixes, additional capabilities and tests. I didn't create #619 so I don't believe I can update it or delete it, hence the new PR.

francisdb commented:
Please keep the `AwsS3IntegrationSpec` as the same previous refactor to `S3IntegrationSpec` was no longer doing proper validation of all AWS S3 cases.

I personally think it's better to have a specific test for each service as they each have their edge cases and url path issues.

easel commented:
The AWSS3IntegrationSpec is still there, I just lifted the test methods up to a trait so I could inject the differing config for this patch. The tests are all identical to before and they all pass against minio -- as far as I am concerned if a provider claims to be s3 compatible it should pass the same integration suite.

francisdb commented:
Seems I was a bit too fast, it's the `S3NoMock` test that is totally broken and that was why I re-introduced an old version as the AWSS3IntegrationSpec.

francisdb commented on s3/src/main/resources/reference.conf:
```diff
@@ -73,4 +73,7 @@ akka.stream.alpakka.s3 {
   path-style-access = false
+
+  # Custom endpoint url, used for alternate s3 implementations
+  endpoint-url = null
```
I wonder if this is discouraged

> If you do need a setting to be optional: checking hasPath() in advance should be the same amount of code (in Java) as checking for null afterward, without the risk of NullPointerException when you forget. In Scala, you could write an enrichment class like this to use the idiomatic Option syntax:
https://github.com/lightbend/config#how-to-handle-defaults

easel replied:
Perhaps? I guess the the alternate solution would be to include the key, commented out. It seems like it should be documented in the reference.conf file in some capacity or another.

francisdb commented on s3/src/test/java/akka/stream/alpakka/s3/javadsl/DocSnippets.java:
```diff
@@ -46,7 +47,8 @@ public String getRegion() {
                 regionProvider,
-                true
+                true,
+                Option.<String>none().asScala() //FIXME
```
Still needs fix to allow a java Optional?

easel replied:
I'm not sure I'm following. Are you suggesting we could simplify that last line?

francisdb replied:
I saw the FIXME so my guess was you wanted something cleaner?

easel replied:
Ah, no, sorry, I carried that over from master when I merged this branch with master. I'm not sure what the intent was... off-hand I'd guess they'd prefer to provide something smarter than a hardcoded none but weren't sure what.

easel commented:
Yeah, I couldn't figure that one out at all =p The AWSS3IntegrationSpec worked ok, I just added some delete calls to make it idempotent.

francisdb commented on s3/src/test/java/akka/stream/alpakka/s3/javadsl/S3ClientTest.java:
```diff
@@ -58,7 +59,8 @@ public String getRegion() throws SdkClientException {
             regionProvider("us-east-1"),
-            false
+            false,
+            akka.japi.Option.<String>none().asScala() //FIXME
```
Still needs fix to allow a java Optional?

ennru replied:
What about `scala.Option.empty()`?
We should add Java API to S3Settings to bridge the case class. (see e.g. JmsSourceSettings), but it doesn't need to be done in the PR.

ennru commented:
@easel and @francisdb can you get this one ready for merge and release? I'd like to get 0.16 out on Thursday.

francisdb commented:
> francisdb approved these changes 5 days ago 
just added some small remarks

takashima0411 commented:
@easel sorry for delay, and thank you for the patches. 
Minio seems good  :+1:

easel commented:
Thanks all for the feedback. I've rebased against master and adjusted things per earlier comments.

- Removed the remaining references to baseUrl
- Default to a commented out endpoint-url key, and added tests that ensure missing keys and resetting to null both are handled as None
- Switched java options to scala.Option.empty(), which compiles. I don't use java api's anymore so I can't say if it's a better solution personally but the code is cleaner!

ennru commented:
Fixes #181 

Thank you for you contribution! Keep going!

easel commented:
Thanks @ennru !

ennru commented:
Cheers. We 'll release 0.16 on Thursday.

l15k4 commented:
Thanks guys, it seems to be working with https://github.com/findify/s3mock but I still cannot figure out why multipart upload doesn't work on https://github.com/jubos/fake-s3 ...

I set it this way : 
```
      endpointUrl = Some(s"http://localhost:$randomPort"),
      pathStyleAccess = true
```

```
akka.http.scaladsl.unmarshalling.Unmarshaller$UnsupportedContentTypeException: Unsupported Content-Type, supported: application/xml, application/octet-stream
	at akka.http.scaladsl.unmarshalling.Unmarshaller$UnsupportedContentTypeException$.apply(Unmarshaller.scala:158)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$EnhancedFromEntityUnmarshaller$.$anonfun$forContentTypes$3(Unmarshaller.scala:114)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$$anon$1.apply(Unmarshaller.scala:58)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$EnhancedUnmarshaller$.$anonfun$mapWithInput$3(Unmarshaller.scala:91)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$$anon$1.apply(Unmarshaller.scala:58)
	at akka.http.scaladsl.unmarshalling.Unmarshaller.$anonfun$transform$3(Unmarshaller.scala:23)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$$anon$1.apply(Unmarshaller.scala:58)
	at akka.http.scaladsl.unmarshalling.Unmarshal.to(Unmarshal.scala:25)
	at akka.stream.alpakka.s3.impl.S3Stream.$anonfun$initiateMultipartUpload$3(S3Stream.scala:216)
```

bandrzejczak commented:
@l15k4 Yes it does. While using standard S3 client I had to disable chunked encoding to make it work with fake-s3. It still allows for sending your document in multiple chunks though, right?
